### ISSUE TITLE
Support for proxy protocol v2

### ISSUE THREAD

**Issue by UltraInstinct14** [loxilb-io/loxilb#675]
**Is your feature request related to a problem? Please describe.**
In many NAT nodes used by loxilb, the original sourceIP gets obfuscated at the endpoint. It might result in problem of accounting various things related to the original source such as abuse, DDOS attacks or to keep simple statistical counts.

**Describe the solution you'd like**
loxilb needs to implement proxy protocol v2. Currently many web-servers already support it

**Describe alternatives you've considered**
N/A

**Additional context**
https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt

**Comment by barryiron** [comment]
Also in need of this feature, would be great if it can be prioritized. Thanks!

**Comment by TrekkieCoder** [comment]
Landed initial support. Details can be found [here](https://docs.loxilb.io/latest/proxy-protocol-v2/)

**Comment by 6547709** [comment]
I just did a test, using Ingress-Nginx, but I can't access ingress.

ingress-nginx logs(10.40.41.91/92 is loxilb'ip):
```
2024/11/27 14:05:00 [error] 578#578: *96633 broken header: "" while reading PROXY protocol, client: 10.40.41.91, server: 0.0.0.0:443
2024/11/27 14:05:00 [error] 862#862: *96700 broken header: "" while reading PROXY protocol, client: 10.40.41.92, server: 0.0.0.0:443
```

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    loxilb.io/staticIP: 10.60.11.10
    loxilb.io/usepodnetwork: "yes"
    loxilb.io/useproxyprotov2: "yes"
    meta.helm.sh/release-name: ingress-nginx
    meta.helm.sh/release-namespace: ingress-nginx
  name: ingress-nginx-controller
  namespace: ingress-nginx
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.253.5.62
  clusterIPs:
  - 10.253.5.62
  externalTrafficPolicy: Local
  healthCheckNodePort: 32261
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  loadBalancerClass: loxilb.io/loxilb
  ports:
  - appProtocol: http
    name: http
    nodePort: 30624
    port: 80
    protocol: TCP
    targetPort: http
  - appProtocol: https
    name: https
    nodePort: 30100
    port: 443
    protocol: TCP
    targetPort: https
  selector:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: ingress-nginx
    app.kubernetes.io/name: ingress-nginx
  sessionAffinity: None
  type: LoadBalancer
```

```
core@LoxiLB-02 ~ $ docker exec -it loxilb loxicmd get lb -o wide
|   EXT IP    | SEC IPS | SOURCES | HOST | PORT | PROTO |                      NAME                      | MARK | SEL  |     MODE     |   ENDPOINT   | EPORT | WEIGHT | STATE  | COUNTERS  |
|-------------|---------|---------|------|------|-------|------------------------------------------------|------|------|--------------|--------------|-------|--------|--------|-----------|
| 10.60.11.10 |         |         |      |   80 | tcp   | ingress-nginx_ingress-nginx-controller:default |    0 | rr   | fullnat:ppv2 | 10.254.2.137 |    80 |      1 | -      | 0:0       |
|             |         |         |      |      |       |                                                |      |      |              | 10.254.5.137 |    80 |      1 | -      | 0:0       |
|             |         |         |      |      |       |                                                |      |      |              | 10.254.6.144 |    80 |      1 | -      | 0:0       |
|             |         |         |      |      |       |                                                |      |      |              | 10.254.1.185 |    80 |      1 | -      | 0:0       |
| 10.60.11.10 |         |         |      |  443 | tcp   | ingress-nginx_ingress-nginx-controller:default |    0 | rr   | fullnat:ppv2 | 10.254.2.137 |   443 |      1 | -      | 86:56498  |
|             |         |         |      |      |       |                                                |      |      |              | 10.254.5.137 |   443 |      1 | -      | 104:66842 |
|             |         |         |      |      |       |                                                |      |      |              | 10.254.6.144 |   443 |      1 | -      | 103:65749 |
|             |         |         |      |      |       |                                                |      |      |              | 10.254.1.185 |   443 |      1 | -      | 88:57036  |

```

```
core@LoxiLB-02 ~ $ docker exec -it loxilb loxicmd get ip
| DEVICE NAME |          IP ADDRESS          |
|-------------|------------------------------|
| docker0     | 172.17.0.1/16                |
| ens192      | 10.40.41.92/24               |
|             | fe80::250:56ff:fe88:1e5c/64  |
| ens224      | 10.60.2.92/16                |
|             | fe80::250:56ff:fe88:93a/64   |
| llb0        | fe80::acbb:f9ff:fef1:d170/64 |
| lo          | 127.0.0.1/8                  |
|             | ::1/128                      |
|             | 10.60.11.7/32                |
|             | 10.60.11.3/32                |
|             | 10.60.11.4/32                |
|             | 10.60.11.5/32                |
|             | 10.60.11.10/32               |
```

**Comment by 6547709** [comment]
ok. I'll look into why it doesn't work.

**Comment by 6547709** [comment]
I have found the cause of the problem. Ingress-Nginx only supports Proxy_Protocol V1. LoxiLB now uses V2, which causes protocol mismatch. I wonder if support for V1 will be added.

**Comment by TrekkieCoder** [comment]
When we implemented v2, it seemed that v1's usage is not widespread but it might be a good addition as well. Will keep this thread posted on this.

**Comment by barryiron** [comment]
FYI, this feature doesn't work when using windows 10 or windows 11 as the client, tested both with curl and chrome. It works fine with both Linux and Mac OS X as the client though. According to tcpdump the checksum is incorrect on the first tcp data packet and also the proxy header looks a bit broken.

**Comment by syserr0r** [comment]
Is there anything that I can do or provide to assist with this issue?

**Comment by xzc949065627-stack** [comment]
I’ve encountered the same problem. When using the fullnat:ppv2 mode, Linux can access it normally, but Windows 11 cannot.

**Comment by Eric-max-tach** [comment]
@inhogog2  I have the same problem on Windows. This bug can be fixed on Windows?

**Comment by inhogog2** [comment]
Hi @Eric-max-tach, Thanks for the heads-up. 

Hi @barryiron, @syserr0r, @xzc949065627-stack, Sorry to late follow this issue. We will push a fix soon.

**Comment by backguynn** [comment]
@Eric-max-tach @barryiron @syserr0r @xzc949065627-stack

The root cause is in how loxilb inserted the PROXY v2 header: it was spliced
inline into the first TCP data packet (the TLS ClientHello). If that packet
gets GRO-merged into a GSO super-packet at loxilb's NIC, the insertion
corrupts the byte stream (the header bytes vanish, leaving a 28-byte sequence
hole — matching the "incorrect checksum / broken proxy header" tcpdump
observations earlier in this thread). Windows clients' first-flight traffic
patterns tend to trigger that GRO merge, while Linux/macOS ClientHellos
usually arrive as a single segment — which also explains why the same Windows
machine worked through WSL.

This is now fixed: the header is inserted on the handshake ACK instead (never
a GSO super-packet). Fix merged in loxilb-ebpf (loxilb-io/loxilb-ebpf#87),
loxilb bump + regression test in #1092. Once that lands in the `latest`
image, please re-test from your Windows clients — feedback welcome.

Until then, a workaround that should unblock Windows clients immediately:
disable GRO/LRO on loxilb's data-plane interfaces
(`ethtool -K <if> gro off lro off`).

**Comment by gengafdafd** [comment]
@backguynn  @inhogog2    Hello, we are still having issues with the verification.

We upgraded loxilb to a newer version and repeated the validation, expecting the Windows client issue to be resolved. However, the issue still persists.

Linux clients work normally, but Windows browser/Xsync traffic continues to fail when using `fullnat` with PROXY protocol v2.

The TCP handshake succeeds, but the connection stalls during the initial data exchange. The backend nginx reports errors while reading the PROXY protocol or during the TLS handshake. Disabling TSO/GSO and checksum offload on the loxilb node did not resolve the issue.

This suggests that the issue may still be related to the `fullnat + PROXY protocol v2` eBPF datapath, possibly involving Windows-specific TCP segmentation, packet ordering, or initial payload handling.


## Environment

- loxilb image: `loxilb:v0.9.8.8`
- loxilb mode: `fullnat` (`mode=2`)
- PROXY protocol v2: enabled
- Host kernel: `5.14.0-570.12.1.el9_6.x86_64`
- Client types:
  - Linux client: works
  - Windows browser/Xsync client: fails
- Backend: nginx HTTPS

## Topology

```text
Windows/Linux client
        |
        v
VIP: 10.122.128.217:8081
        |
        | loxilb fullnat + PPv2
        v
Backend: 10.122.131.141:443
```

Problem
Linux clients can access the VIP successfully.
Windows browser/Xsync clients cannot access:


https://10.122.128.217:8081/
The TCP handshake completes, but the connection stalls after the first client data packets. The backend repeatedly sends ACKs with Ack=1 and SACKs the subsequent data.

Example:
10.122.128.217:50479 > 10.122.131.141:443: [SYN]
10.122.131.141:443 > 10.122.128.217:50479: [SYN, ACK]
10.122.128.217:50479 > 10.122.131.141:443: seq 1:29, length 28
10.122.128.217:50479 > 10.122.131.141:443: seq 29:463, length 434
10.122.131.141:443 > 10.122.128.217:50479: ACK 1, SACK 29:463
The later client data is retransmitted repeatedly, while the backend does not advance the cumulative ACK.

Nginx reports errors such as:
broken header while reading PROXY protocol

recv() failed (104: Connection reset by peer) while SSL handshaking
The application also reports:

text


Xsync server returned error (-49977)
Comparison
The same backend works with Linux clients through the same VIP and loxilb rule.

The failure is specific to Windows browser/Xsync traffic. Windows sends a larger and more complex first request/data pattern than the Linux curl test.

Additional testing
We temporarily disabled NIC offload features on the loxilb node:
```
TSO off
GSO off
GRO off
TX checksum off
The problem still reproduced.
```
The offload settings were restored afterwards.

Therefore, this does not appear to be caused only by TSO/GSO/TX checksum offload.

Expected behavior
For a fullnat + proxyprotocolv2 service, loxilb should forward:
valid PROXY protocol v2 header

original client TLS/HTTP payload
The backend nginx should be able to parse the PPv2 header and continue with TLS/HTTP processing.

Actual behavior
Linux traffic works, but Windows traffic stalls after the first data packets.

The backend does not advance the TCP cumulative ACK beyond the initial sequence position and only SACKs later data. Nginx remains in PROXY protocol or TLS handshake processing and eventually reports a broken header, timeout, or connection reset.

**Comment by backguynn** [comment]
@gengafdafd Thanks — that capture is enough to place it.
v0.9.8.8 predates two later fixes; the July one only covered the GSO case.
* loxilb-ebpf#91: dp_ins_ppv2() never wrote the header when doff == 20 (no TCP options). The room was added and lengths adjusted, so the stream carried 28 bytes of uninitialised data instead of a PROXY v2 header — nginx's broken header. Windows omits the timestamp option by default; Linux sends it (doff == 32) and lands on a working branch. Hence Windows-only.
* loxilb-ebpf#93: the ppv2 path emitted an invalid TCP checksum whenever loxilb's ingress skb is CHECKSUM_PARTIAL (containerised loxilb, veth/tun peers). Your ACK 1 / SACK 29:463 is that signature — the backend drops the 28-byte segment and SACKs the rest. ethtool -K on the NIC does not avoid it, which is why your offload test didn't help.
Both are on main (loxilb #1111);    

no tagged release has them yet. Please retest with `ghcr.io/loxilb-io/loxilb:latest`.
